### PR TITLE
APEXCORE-740 Load the plugins from setup of setup plugin manager instead of constructor

### PR DESCRIPTION
**Pull request by chinmaykolhatkar**
RCA given in the jira.

### CODE REVIEW

**Comment by chinmaykolhatkar** [comment]
@tushargosavi Can you please review and merge?

**Comment by chinmaykolhatkar** [comment]
@PramodSSImmaneni @tushargosavi Can you please review it.

**Comment by pramodin** [comment]
Looking into it

File: engine/src/main/java/com/datatorrent/stram/client/StramAppLauncher.java [hunk 1]
```diff
@@ -535,7 +535,8 @@ public void runLocal(AppFactory appConfig) throws Exception
 
   public URLClassLoader loadDependencies()
   {
-    URLClassLoader cl = URLClassLoader.newInstance(launchDependencies.toArray(new URL[launchDependencies.size()]));
+    ClassLoader currentCl = Thread.currentThread().getContextClassLoader();
+    URLClassLoader cl = URLClassLoader.newInstance(launchDependencies.toArray(new URL[launchDependencies.size()]), currentCl);
```
Comment (pramodin) [inline engine/src/main/java/com/datatorrent/stram/client/StramAppLauncher.java]: Is this needed if the necessary artifacts are available in the launch dependencies

Comment (chinmaykolhatkar) [reply engine/src/main/java/com/datatorrent/stram/client/StramAppLauncher.java]: Its not necessary. Made the changes to load the plugin with the same classloader.

File: engine/src/main/java/com/datatorrent/stram/client/StramAppLauncher.java [hunk 2]
```diff
@@ -580,6 +580,7 @@ public ApplicationId launchApp(AppFactory appConfig) throws Exception
     loadDependencies();
     Configuration conf = propertiesBuilder.conf;
     conf.setEnum(StreamingApplication.ENVIRONMENT, StreamingApplication.Environment.CLUSTER);
+    propertiesBuilder.loadPlugins();
```
Comment (pramodin) [inline engine/src/main/java/com/datatorrent/stram/client/StramAppLauncher.java]: This misses other ways in which application can be launched such as local mode or launch from api.

Comment (pramodin) [reply engine/src/main/java/com/datatorrent/stram/client/StramAppLauncher.java]: How about doing this in the setup method of the DAGSetupPluginManager one time.

Comment (chinmaykolhatkar) [reply engine/src/main/java/com/datatorrent/stram/client/StramAppLauncher.java]: Yes.. Good point and suggestion sounds fine. I'll make the changes.

**Comment by chinmaykolhatkar** [comment]
@PramodSSImmaneni Made the changes. Now there is a single file change in DAGSetupPluginManager. The plugins are loaded in setup one time. Also rebased and squashed the PR.

File: engine/src/main/java/com/datatorrent/stram/plan/logical/DAGSetupPluginManager.java [hunk 3]
```diff
@@ -52,9 +52,13 @@
 
   private final Table<DAGSetupEvent.Type, DAGSetupPlugin, EventHandler<DAGSetupEvent>> table = HashBasedTable.create();
 
-  private void loadVisitors(Configuration conf)
+  public DAGSetupPluginManager(Configuration conf)
   {
     this.conf = conf;
+  }
+
+  public void loadVisitors()
```
Comment (pramodin) [inline engine/src/main/java/com/datatorrent/stram/plan/logical/DAGSetupPluginManager.java]: Why public?

**Comment by pramodin** [comment]
Could you also change the title of the PR.

**Comment by chinmaykolhatkar** [comment]
@PramodSSImmaneni Done with the changes. Can you please review and merge?

**Comment by pramodin** [comment]
@chinmaykolhatkar can you rebase

**Comment by chinmaykolhatkar** [comment]
@PramodSSImmaneni Rebase done.

File: engine/src/main/java/com/datatorrent/stram/plan/logical/DAGSetupPluginManager.java [hunk 4]
```diff
@@ -52,9 +52,13 @@
 
   private final Table<DAGSetupEvent.Type, DAGSetupPlugin, EventHandler<DAGSetupEvent>> table = HashBasedTable.create();
 
-  private void loadVisitors(Configuration conf)
+  public DAGSetupPluginManager(Configuration conf)
```
Comment (pramodin) [inline engine/src/main/java/com/datatorrent/stram/plan/logical/DAGSetupPluginManager.java]: Should this be private?

File: engine/src/main/java/com/datatorrent/stram/plan/logical/DAGSetupPluginManager.java [hunk 5]
```diff
@@ -145,8 +150,7 @@ public void dispatch(DAGSetupEvent event)
 
   public static synchronized DAGSetupPluginManager getInstance(Configuration conf)
```
Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/plan/logical/DAGSetupPluginManager.java:151]: @tushargosavi @chinmaykolhatkar @PramodSSImmaneni Does the method need to be synchronized?

**Pull request merged**